### PR TITLE
Smoke me/khw locale

### PR DESCRIPTION
**Pull request by khwilliamson**

### CODE REVIEW

**Comment by khwilliamson** [comment]
@tonycoz does this new commit address your concern?  Do I need something in makedef.pl?

File: ext/POSIX/POSIX.xs [hunk 1]
```diff
@@ -3348,30 +3346,49 @@ write(fd, buffer, nbytes)
...
-        LOCALE_UNLOCK;
+        errno = 0;
+        if (! SvOK(s)) {
+#ifdef USE_MBRLEN
+            /* Initialize the shift state in PL_mbrlen_ps.  It's a good guess
+             * that the initial state is all zeros.  On some platforms, it is
+             * documented that pointing the first parameter to L'\0'
+             * initializes the state; on others, that doesn't work, but making
+             * the parameter NULL does.  Perhaps NULL works on all platforms,
+             * but this seems safer. */
```
Comment (tonycoz) [inline ext/POSIX/POSIX.xs:3369]: This is (badly) documented in the C standard as

> A zero-valued mbstate_t object is (at least) one way to describe an initial conversion state. A zerovalued
mbstate_t object can be used to initiate conversion involving any multibyte character sequence, in any LC_CTYPE category setting.

but doesn't define what "zero-valued" is for what is to us an opaque object.  It's up to you whether this comment should reflect that

File: ext/POSIX/POSIX.xs [hunk 2]
```diff
@@ -3348,30 +3346,49 @@ write(fd, buffer, nbytes)
...
-        RETVAL = mbrlen(s, n, &ps); /* Prefer reentrant version */
-#else
-        /* This might prevent some races, but locales can be switched out
-         * without locking, so this isn't a cure all */
-        LOCALE_LOCK;
-
-        RETVAL = mblen(s, n);
-        LOCALE_UNLOCK;
+        errno = 0;
+        if (! SvOK(s)) {
```
Comment (tonycoz) [inline ext/POSIX/POSIX.xs:3362]: You need SvGETMAGIC(s) before checking SvOK(), and with that you *may* want to use SvPV_nomg() (except, see below where I talk about byte vs character strings).

File: ext/POSIX/POSIX.xs [hunk 3]
```diff
@@ -3348,30 +3346,49 @@ write(fd, buffer, nbytes)
...
+            RETVAL = 0;
+#else
+            LOCALE_LOCK;
+            RETVAL = mblen(NULL, n);
+            LOCALE_UNLOCK;
 #endif
+        }
+        else {  /* Not resetting state */
+            size_t len;
+            char * string = SvPV(s, len);
```
Comment (tonycoz) [inline ext/POSIX/POSIX.xs:3380]: Is `s` a character or byte string?
````
$ ./perl -Ilib unitest.pl
4
2
equal
$ cat unitest.pl
use POSIX qw(mblen setlocale LC_ALL);
use locale;

setlocale(LC_ALL, "C.UTF-8")
  or die;

my $x = "\xf0\x90\xbf\xb0";

mblen(undef);
print mblen($x), "\n";

utf8::upgrade(my $y = $x);

print mblen($y), "\n";

print $x eq $y ? "equal" : "not equal", "\n";
````
This isn't a problem you've introduced, but it's still a problem.

Should this be using SvPVbyte(_nomg)() ?

**Pull request closed without merging**